contracts/libraries/DelegateGuard.sol

```diff
@@ -0,0 +1,79 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.17;
+
+library LibBitSlot {
+  function get(bytes32 slot, uint8 index) internal view returns (bool isSet) {
+    assembly {
+      isSet := and(shr(index, sload(slot)), 1)
+    }
+  }
+
+  function set(
+    bytes32 slot,
+    uint8 index,
+    bool isSet
+  ) internal {
+    assembly {
+      let value := sload(slot)
+      let shift := and(index, 0xff)
+      // Isolate the bit at `shift`.
+      let bit := and(shr(shift, value), 1)
+      // Xor it with `shouldSet`. Results in 1 if both are different, else 0.
+      bit := xor(bit, isSet)
+      // Shifts the bit back. Then, xor with value.
+      // Only the bit at `shift` will be flipped if they differ.
+      // Every other bit will stay the same, as they are xor'ed with zeroes.
+      sstore(slot, xor(value, shl(shift, bit)))
+    }
+  }
+}
+
+abstract contract Slot {
+  function _slot() internal pure virtual returns (bytes32);
+}
+
+abstract contract DelegateGuard is Slot {
+  error AlreadyInitialized();
+  error CallTypeRestricted();
+
+  using LibBitSlot for bytes32;
+
+  uint256 internal constant _ORIGINAL_BIT_INDEX = 96;
+  uint256 internal constant _INITIALIZED_BIT_INDEX = 95;
+
+  modifier restrictDelegate(bool useDelegate) virtual {
```

Comment:
I think the value of the bool var should be as the same as the name, not in reverted value, to avoid misuse.

```suggestion
  modifier restrictDelegate(bool _shouldRestrict) virtual {
```

Reply:
Please rename this var, and revert the corresponding, e.g.

```
if iszero(xor(eq(original, address()), useDelegate_)) {
```

is modified to

```
if iszero(xor(not(eq(original, address()), useDelegate_))) {
```

One more `not` opcode is ok.

Reply:
I have commented the logic, please check out.